### PR TITLE
Automate switch of data fraction used in processing

### PR DESCRIPTION
**Pull request by darafferty**
This PR allows the fraction of data that is processed to be changed for the final iteration (e.g., so that one can do selfcal with 20% of the data and then automatically do a final pass with 100% of the data).

### CODE REVIEW

File: rapthor/process.py [hunk 1]
```diff
@@ -77,7 +77,37 @@ def run(parset_file, logging_level='info'):
                 if has_diverged:
                     log.warning("Selfcal has diverged (ratio of current image noise "
                                 "to previous value is > {})".format(field.divergence_ratio))
-                log.info("Stopping at iteration {0} of {1}".format(index+1, len(strategy_steps)))
+                log.info("Stopping selfcal at iteration {0} of {1}".format(index+1, len(strategy_steps)))
                 break
 
+    # Run with the final data fraction if needed
+    if parset['final_data_fraction'] != parset['data_fraction']:
```
Comment (gmloose) [inline rapthor/process.py:84]: This is comparing floats isn't it. Maybe use `np.isclose` or something similar.

File: rapthor/process.py [hunk 2]
```diff
@@ -77,7 +77,37 @@ def run(parset_file, logging_level='info'):
...
+        log.info("Starting final iteration with a data fraction of "
+                 "{0:.2f}".format(parset['final_data_fraction']))
+
+        # Set peel_outliers to that of initial iteration, since the observations
+        # will be regenerated and outliers may need to be peeled
+        step['peel_outliers'] = strategy_steps[0]['peel_outliers']
+        field.update(step, index+2, final=True)
```
Comment (gmloose) [inline rapthor/process.py:91]: I don't know exactly why you need to do `index+2` here. Maybe a small comment explaining this?

**Comment by gmloose** [comment]
One question I have: what happens if `final_data_fraction` is larger than `data_fraction`?

**Comment by gmloose** [comment]
As a general remark, not related to this MR, it would probably be wise to do a bit more input checking and/or sanitization when reading the parset file. For example, there's currently no check that `0 <= data_fraction <=1`.

**Comment by darafferty** [comment]
> One question I have: what happens if `final_data_fraction` is larger than `data_fraction`?

Good question -- it should work fine, but would be a strange thing to do (at least, I can't think of a good reason to do it). Perhaps it would be good to add a warning for this case.

File: rapthor/process.py (hunk 1 above)
Comment (darafferty) [reply rapthor/process.py]: Good idea. I updated it to use `np.isclose`.

File: rapthor/process.py (hunk 2 above)
Comment (darafferty) [reply rapthor/process.py:95]: Yep, I added a comment to make it more clear.

**Comment by gmloose** [comment]
> Edit: actually, I read your comment as: `final_data_fraction` is smaller than `data_fraction`. Was that what you meant?

No, I really meant what happens if `final_data_fraction > data_fraction`? 
If it doesn't make sense for `final_data_fraction` to be either larger or smaller than `data_fraction`, then there's no need to be able to set it at all, right?

**Comment by darafferty** [comment]
Ah, I see. Normally, the final fraction should be larger than the initial one, as selfcal (which is done with the initial fraction) generally doesn't need the full-sensitivity images to work well (since the brightest sources dominate), so you can save time by processing just 20% or so. But often you want the final image to be as deep as possible, so after selfcal is done the full dataset should be processed. I would expect either `final_data_fraction = 1` or `final_data_fraction = data_fraction` almost all of the time, but perhaps there are cases where something else would be useful.

**Comment by gmloose** [comment]
So, `data_fraction` refers to the initial fraction? If that's the case, maybe it should be renamed to `initial_data_fraction` then; or maybe they should be called `calibration_data_fraction` and `imaging_data_fraction` instead. Just some ideas.

**Comment by darafferty** [comment]
That's a good point -- how about `selfcal_data_fraction`? I can change it to that, but also keep `data_fraction` for now (and warn that it is deprecated).

**Comment by gmloose** [comment]
Yeah, `selfcal_data_fraction` is fine with me as well. Do you agree with `imaging_data_fraction`?

**Comment by darafferty** [comment]
I think I prefer to keep it as `final_data_fraction`, if that's alright with you, since the final pass includes calibration as well as imaging.

**Comment by gmloose** [comment]
Sure, no problem.

**Pull request merged**